### PR TITLE
Sentry check "tilt mode"

### PR DESCRIPTION
**Pull request by jden**
One of the symptoms during a sentry collapse is increased latency
in the Node.js eventloop, which blocks over events like network IO.
This causes sentrys to be late sending their keepalives, but it also
causes the sentry system to be late checking the health of remote
sentries in the cluster. If process A's latency exceeds the sentry
healthcheck interval, it may result in every other server in the
cluster being reported as down, triggering many false client
disconnects and a flood of radar messages and redis commands.

This PR adds a check for clock skew, ensuring that the sentry check
happens no more than 1000ms after it was scheduled, otherwise it skips
checking on this interval and waits for the next (30s later).

The idea is borrowed from Redis' Sentinel system http://redis.io/topics/sentinel#tilt-mode

Fun fact, the name "tilt mode" comes from arcade pinball machines:
to discourage cheating, they detect if there is an unstable environment
caused by the player bumping the table. If there is, it knows that the
physical state of the game cannot be trusted, so the controls seize
temporarily. Vending machines implement similar protections.

References
========
- [RADAR-638](https://zendesk.atlassian.net/browse/RADAR-638)

Required
========
- [ ] :+1: from @zendesk/radar

Risks
========
- None

### CODE REVIEW

File: server/sentry.js [hunk 1]
```diff
@@ -241,11 +241,26 @@ Sentry.prototype._refresh = function () {
...
+    this._loadAndCleanUpSentries()
+  }
+
+  this._lastChecked = Date.now()
   this._checkSentriesTimer = setTimeout(this._checkSentries.bind(this), interval)
 }
 
+// tilt mode: based on http://redis.io/topics/sentinel#tilt-mode
+// before recognizing a remote sentry as down,
+// make sure we trust our own observations by checking for clock skew.
```
Comment (lucasefe) [inline server/sentry.js]: I love this: "make sure we trust our own observations by checking for clock skew"

**Comment by vanchi-zendesk** [comment]
this is a bit weird, since a heavily lagged server can go without checking for a long time. Even a moderate lag can do bad things here. We should eventually check the sentries or we might miss out of gone servers. Let's discuss this further offline.

**Comment by vanchi-zendesk** [comment]
One good way to deal with this would be to adjust our own keep-alive expiry based on the lag we are seeing. If we see high lag, we double our own expiry when we update it. Another thing we can do is : increase our sentry keep-alive frequency as well.

**Comment by jden** [comment]
@vanchi-zendesk the problem is not with the keep-alive not being updated: a server that hasn't sent a keep-alive in more than X seconds should rightly be considered down. The problem is when a lagged process is late in checking for other expired sentries.

> a heavily lagged server can go without checking for a long time

This is desirable behavior, because a false positive for sentry down creates additional load on that already lagged process (S), and on the cluster. Delaying a check, on the other hand, results in a false negative, i.e., a remote server (R) is observed to be up when it is actually down. Delaying the sentry down event delays handling implicit disconnects for R's clients  only for presence resources managed on S.

**Comment by vanchi-zendesk** [comment]
When a server has not updated sentry within expected time, the other servers may already have marked this server down. With this, it's possible that as load increases, no server essentially ever checks for sentry at all. Although, I understand the point about not having to deal with failures when heavily loaded, it remains essential to how presence works. I would rather fail due to heavy load, rather than keep offline clients showing up online.

"The problem is when a lagged process is late in checking for other expired sentries." This is a bigger problem that we realize: https://github.com/zendesk/radar/pull/220/files#diff-7c4383a7865fffe451650ef51b299dffL90
Notice that we check the sentry expiration from locally stored values. So we expect ourselves to consistently check the latest values from Redis here. If we expand the expiryOffsetFromNow() value for others and ourselves based on our own lag, we have a better chance.

**Comment by jden** [comment]
>  I would rather fail due to heavy load, rather than keep offline clients showing up online.

Okay, in that case I'll close the PR.

**Pull request closed without merging**